propra_webscience_ws24/data/data_preprocessing.py

```diff
@@ -0,0 +1,239 @@
+            "stopwords_strategy={} in {}.",
+            split,
+            normalization_strategy.value,
+            stopwords_removal_strategy.value,
+            humanize.precisedelta(elapsed_time_delta, minimum_unit="milliseconds"),
+        )
+
+
+def _get_normalization_function(
+    normlization_strategy: TextNormalizationStrategy,
+) -> Callable[[str], str]:
+    text_normalization_functions = {
+        TextNormalizationStrategy.LEMMATIZER: WordNetLemmatizer().lemmatize,
+        TextNormalizationStrategy.PORTER: PorterStemmer().stem,
+        TextNormalizationStrategy.NONE: lambda x: x,
+    }
+
+    return text_normalization_functions[normlization_strategy]
+
+
+def _normalize_text(
+    text: str,
+    normalization_strategy: TextNormalizationStrategy,
+    stopwords_removal_strategy: StopwordRemovalStrategy,
+) -> str:
+    tweet_tokenizer = TweetTokenizer(preserve_case=False, strip_handles=True, reduce_len=True)
+    tokens = tweet_tokenizer.tokenize(text)
+
+    text_normalization_function = _get_normalization_function(normalization_strategy)
+    normalized_tokens = [text_normalization_function(token) for token in tokens]
+
+    if stopwords_removal_strategy == StopwordRemovalStrategy.KEEP:
+        pass
+    elif stopwords_removal_strategy == StopwordRemovalStrategy.REMOVE_DEFAULT_NLTK:
+        normalized_tokens = [
+            t for t in normalized_tokens if t not in DEFAULT_NLTK_STOPWORDS
+        ]
+    elif stopwords_removal_strategy == StopwordRemovalStrategy.REMOVE_CUSTOM:
+        normalized_tokens = [
+            t for t in normalized_tokens if t not in _get_custom_stopwords()
+        ]
+    else:
+        raise NotImplementedError(
+            f"Stopword removal strategy "
+            f"{stopwords_removal_strategy} is not implemented."
+        )
+
+    return " ".join(normalized_tokens)
+
+
+@lru_cache(maxsize=1)
+def _get_custom_stopwords():
+    """
+    Get a set of custom stopwords that explicitly handles contractions and negations.
+    """
+    global _CUSTOM_STOPWORDS
+    if _CUSTOM_STOPWORDS is None:
+        stop_words = set(stopwords.words("english"))
+
+        negation_words = {
+            "no",
+            "not",
+            "nor",
+            "neither",
+            "never",
+            "none",
+            "don't",
+            "doesn't",
+            "didn't",
+            "haven't",
+            "hasn't",
+            "won't",
+            "wouldn't",
+            "can't",
+            "couldn't",
+            "shouldn't",
+            "weren't",
+            "wasn't",
+            "isn't",
+            "aren't",
+            "ain't",
+            "mightn't",
+            "mustn't",
+            "needn't",
+            "shan't",
+            "oughtn't",
+            "daren't",
+            "hadn't",
+        }
+        _CUSTOM_STOPWORDS = {
+            word
+            for word in stop_words
+            if not any([negation in word for negation in negation_words])
```

Comment:
Hier würde `if word not in negation_words` ausreichen?

Reply:
Ja, stimmt :+1:. Geht sogar noch übersichtlicher mit `stop_words - negation_words`.

Reply:
Das geht so einfach? Wild